p4-16/psa/PSA.mdk

```diff
@@ -523,34 +523,69 @@ extern Checksum<W> {
-packets that trigger a table in a Match Action unit.
+Counters are a mechanism for keeping statistics.  The control plane
+can read counter values.  A P4 program cannot read counter values,
+only update them.  If you wish to implement a feature involving
+sequence numbers, for example, use Registers instead (Section
+[#sec-registers]).
+
+Direct counters are counters associated with a particular P4 table.
+The primary differences between direct counters and indirect counters
+are:
+
+- Number of independently updatable counter values:
+  - A single instantiation of a direct counter always contains as many
+    independent counter values as the table with which it is
+    associated.
+  - You must specify the number of independent counter values for an
+    indirect counter when instantiating it.  This number of counters
+    need not be the same as the size of any P4 table.
+- Where counter updates are allowed in the P4 program:
+  - For a direct counter, you may only invoke its `count` method from
+    inside the actions of the table with which it is associated, and
+    this always updates the counter value associated with the matching
+    table entry.
+  - For an indirect counter, you may invoke its `count` method
+    anywhere in the P4 program where extern object method invocations
+    are permitted (e.g. inside actions, or directly inside a control's
+    `apply` block), and every such invocation must specify the index
+    of the counter value to be updated.
+
+One may implement packet counters by always using the `count` method
+that adds 1 to the counter value.
+
+One may implement byte counters by using the `count` method that adds
+a caller-specified `add_value` to the counter, containing a value that
+the P4 program wishes to use for the length of the packet.
+
+We mention these as common uses cases for counters.  You may of course
+use them to count things other than packets and bytes.
 
-Direct counters fire when the count method is invoked in an action, and
-have an instance for each entry in the table.
-
-
-### Counter types
-
-```
-enum CounterType_t {
-    packets,
-    bytes,
-    packets_and_bytes
-}
-```
 
 ### Counter
 
 ```
-extern Counter<W, S> {
-  Counter(S n_counters, W size_in_bits, CounterType_t counter_type);
-  void count(in S index, in W increment);
+/// Indirect counter with n_counters independent counter values, where
+/// every counter value has a size specified by type W.  Each counter
+/// can be updated either by adding 1 via the `count(index)` method,
+/// or by adding a value add_value with size specified by type V.
+/// Type W must contain at least as many bits as type V.
+///
+/// Type N is separate from type S to handle the case when the number
+/// of counter values is a power of 2, but you wish to use the fewest
+/// bits possible in type S for the index.  Type S must contain enough
+/// bits to specify all values in the range [0, n_counters - 1].
```

Comment:
I could fairly easily get rid of the type parameter N by making the type of n_counters bit<32> instead, which with the current p4c then requires casting the n_counters parameter value to bit<32> when you call the constructor.  Not sure if you consider that an improvement, but I've committed such a change to this PR for you to examine.  If you have more ideas for removing more of them, please give specific suggestions, preferably ones that compile with today's compiler, or a version of it planned to be committed in the next couple of months.